torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -591,6 +593,17 @@ ProcessGroupNCCL::ProcessGroupNCCL(
             << "\nUSE_HIGH_PRIORITY_STREAM: "
             << options_->is_high_priority_stream
             << "\nNCCL_DEBUG: " << ncclDebugLevel;
+
+
+  static std::once_flag initialize_ucc_lib_flag;
+  std::call_once(initialize_ucc_lib_flag, [&]{
+    ucc_lib_ = loadTorchUCC();
```

Comment:
Should we just load UCC when CUDA/GPU is not available?

Reply:
One question about UCC dependence. Does OSS UCC depend on NCCL? If so, when UCC is loaded through torch_ucc, is it possible that NCCL will also be loaded? Will this end up with two different versions of NCCL being loaded into PyTorch?

Reply:
Yes, UCC supports using NCCL (as well as UCX and other libraries) as its transport. But in our case, we are not using this functionality. When we use UCC, we should be using the UCX backend and when we want to use NCCL, we will use it directly.

For the question of having multiple versions of NCCL, my understanding is:

## If both PyTorch and UCC are dynamically linked to NCCL

The documentation of `dlopen`[1] says:
> Symbol references in the shared object are resolved using (in order): symbols in the link map of objects loaded for the main program and its dependencies; symbols in shared objects (and their dependencies) that were previously opened with dlopen() using the RTLD_GLOBAL flag; and definitions in the shared object itself (and any dependencies that were loaded for that object).

Because PyTorch is always loaded earlier than torch_ucc, by the time torch_ucc is loaded, symbols of NCCL should have already been available in the main program. So `dlopen` will just use these symbols for UCC.

## If both PyTorch and UCC are statically linked to NCCL

For static linking, symbol resolution should happen at link time. So PyTorch will have its copy of NCCL code, and UCC will also have its copy. Each will use its own copy. The NCCL symbols in `libucc.so` will not be added to the main program for later symbol resolution because we `dlopen` it with `RTLD_LOCAL`, and according to the documentation[1]:
> This is the converse of RTLD_GLOBAL, and the default if neither flag is specified.  Symbols defined in this shared object are not made available to resolve references in subsequently loaded shared objects.

## If PyTorch is statically linked to NCCL and UCC is dynamically linked to NCCL

For this case, PyTorch will use the NCCL that it is compiled with because symbol resolution happened at link time. I don't know which NCCL UCC will use, because I don't know if NCCL's symbol will be copied to PyTorch and added to the link map. If yes, then UCC will use that one from PyTorch. If not, then UCC will loads its own NCCL and use the symbol there.

## If PyTorch is dynamically linked to NCCL and UCC is statically linked to NCCL

PyTorch is always loaded earlier than `torch_ucc`, so PyTorch will load NCCL from the system and use that NCCL. But UCC will still use its own NCCL because all the NCCL function calls are already resolved when `libucc.so` is generated.

# Reference
- [1] https://man7.org/linux/man-pages/man3/dlopen.3.html